src/v/io/io_queue.h

```diff
@@ -0,0 +1,161 @@
+    io_queue& operator=(const io_queue&) = delete;
+    io_queue(io_queue&&) noexcept = delete;
+    io_queue& operator=(io_queue&&) noexcept = delete;
+    ~io_queue() noexcept = default;
+
+    /**
+     * Starts the I/O queue's background dispatcher.
+     */
+    void start() noexcept;
+
+    /**
+     * Stops the I/O queue's background dispatcher.
+     *
+     * It is recommended that close() be invoked prior to stopping the queue.
+     * While Seastar will close the file handle automatically, ensuring that the
+     * queue is drained before the file handle is closed will avoid errors
+     * within Seastar related to closing file handles with pending I/O.
+     */
+    seastar::future<> stop() noexcept;
+
+    /**
+     * Return the path of the backing file.
+     */
+    [[nodiscard]] const std::filesystem::path& path() const noexcept;
+
+    /**
+     * Open the queue and begin dispatching pending I/O requests.
+     *
+     * Requires that opened() be false. The queue will remain in the closed
+     * state if an exceptional future is returned.
+     */
+    seastar::future<> open() noexcept;
+
+    /**
+     * Returns true if the queue is open, and false otherwise.
+     */
+    bool opened() const noexcept;
+
+    /**
+     * Close the queue and release the underlying file handle.
+     *
+     * Waits until inflight I/Os are complete. Pending I/Os remain in the
+     * pending state until the queue is re-opened.
+     *
+     * Requires that opened() be true.
+     */
+    seastar::future<> close() noexcept;
+
+    /**
+     * Submit a read I/O request.
+     */
+    void submit_read(page&) noexcept;
+
+    /**
+     * Submit a write I/O request.
+     *
+     * If a write request is submitted for a page that had been dispatched but
+     * not yet completed then it will be requeued to be written again.
+     */
+    void submit_write(page&) noexcept;
+
+private:
+    friend testing_details::io_queue_accessor;
+
+    using request_list_type = intrusive_list<page, &page::io_queue_hook>;
+
+    /*
+     * file always represents path on the storage backend. the file pointer is
+     * set via open() and cleared via close(). this may happen any number of
+     * times during the life time of the queue.
+     */
+    persistence* storage_;
+    std::filesystem::path path_;
+    seastar::shared_ptr<persistence::file> file_;
+
+    /*
+     * submitted requests first land on the pending list. they are moved to the
+     * running list when they are dispatched. when a request is completed it is
+     * removed from the running list, and passed to the completion callback.
+     *
+     * operations are dispatched under the ops semaphore which is used to
+     * synchronize completion of operations with pending calls to close().
+     */
+    request_list_type pending_;
+    request_list_type running_;
+    completion_callback_type complete_;
+    seastar::semaphore ops_{seastar::semaphore::max_counter()};
```

Comment:
I'm curious why is it set to `max_counter`?

Reply:
FWIW seastar has a `seastar::rwlock` to encapsulate this pattern in a more clear way.

Reply:
the semaphore use is intentional. even though it is configured as a shared-exclusive lock, we'll later use the semaphore to control the number of outstanding IOs.